nova/scheduler/external.py

```diff
@@ -0,0 +1,113 @@
+to reorder and filter hosts based on additional criteria.
+
+The external scheduler API is expected to take a list of weighed hosts and
+their weights, along with the request specification, and return a reordered
+and filtered list of host names.
+"""
+from oslo_log import log as logging
+import requests
+
+from nova.api.validation import validators
+import nova.conf
+from nova import exception
+from nova.scheduler import utils
+
+CONF = nova.conf.CONF
+LOG = logging.getLogger(__name__)
+
+
+# The expected response schema from the external scheduler api.
+# The response should contain a list of ordered host names.
+response_schema = {
+  "type": "object",
+  "properties": {
+    "hosts": {
+      "type": "array",
+      "items": {
+        "type": "string"
+      }
+    }
+  },
+  "required": ["hosts"],
+  "additionalProperties": False,
+}
+
+
+def call_external_scheduler_api(weighed_hosts, weights, spec_obj):
+    """Reorder and filter hosts using an external scheduler service."""
+    if not weighed_hosts:
+        return weighed_hosts
+    if not (url := CONF.filter_scheduler.external_scheduler_api_url):
+        LOG.debug("External scheduler API is not enabled.")
+        return weighed_hosts
+    timeout = CONF.filter_scheduler.external_scheduler_timeout
+
+    json_data = {
+        "spec": spec_obj.obj_to_primitive(),
+        # Extract some flags from the spec to indicate the type of
+        # request. This will allow the external scheduler to quickly
+        # decide if it wants to handle the request or not.
+        "rebuild": utils.request_is_rebuild(spec_obj),
+        "resize": utils.request_is_resize(spec_obj),
+        "live": utils.request_is_live_migrate(spec_obj),
+        "vmware": not utils.is_non_vmware_spec(spec_obj),
+        # Only provide basic information for the hosts for now.
+        # The external scheduler is expected to fetch statistics
+        # about the hosts separately, so we don't need to pass
+        # them here.
+        "hosts": [
+            {
+                "host": h.host,  # e.g. nova-compute-bb123
+                "hypervisor_hostname": h.hypervisor_hostname,
+                "status": h.status,
+            } for h in weighed_hosts
+        ],
+        # Also pass previous weights from the Nova weigher pipeline.
+        # The external scheduler api is expected to take these weights
+        # into account if provided.
+        "weights": weights,
+    }
+    LOG.debug("Calling external scheduler API with %s", json_data)
+    try:
+        response = requests.post(url, json=json_data, timeout=timeout)
+        response.raise_for_status()
+    except requests.RequestException as e:
+        LOG.error("Failed to call external scheduler API: %s", e)
+        return weighed_hosts
+
+    # The external scheduler api is expected to return a json with
+    # a sorted list of host names. Note that no weights are returned.
+    response_json = response.json()
```

Comment:
Could there be a case where the external scheduler returns a HTTP 2XX but not JSON? If yes, this will raise an exception and prohibit scheduling.